File: algorithms/152-maximum-product-subarray/maximum-product-subarray.go
package problem152

// dynamic programming
func maxProduct(nums []int) int {
	prevMin := nums[0]
	prevMax := nums[0]
	result := nums[0]

	for _, n := range nums[1:] {
		tmpMin := n * prevMin
		tmpMax := n * prevMax
		curMin := min(min(n, tmpMax), tmpMin)
		curMax := max(max(n, tmpMax), tmpMin)
		result = max(result, curMax)
		prevMax = curMax
		prevMin = curMin
	}

	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

File: algorithms/152-maximum-product-subarray/maximum-product-subarray.py
"""
152. maximum product subarray

乘积最大子数组
"""

def maxProduct(self, nums: List[int]) -> int:
    """
    dynamic programming
    """
    prev_min = nums[0]
    prev_max = nums[0]
    result = nums[0]

    for n in nums[1:]:
        tmp_min = n * prev_min
        tmp_max = n * prev_max
        cur_max = max(max(n, tmp_min), tmp_max)
        cur_min = min(min(n, tmp_min), tmp_max)
        result = max(result, cur_max)
        prev_max = cur_max
        prev_min = cur_min
    
    return result